x/poolmanager/router_test.go

```diff
@@ -1530,3 +1588,419 @@ func (suite *KeeperTestSuite) setupPools(poolType types.PoolType, poolDefaultSwa
 		return
 	}
 }
+
+func (suite *KeeperTestSuite) TestSplitRouteExactAmountIn() {
+	var (
+		defaultSingleRouteOneHop = []types.SwapAmountInSplitRoute{
+			{
+				Pools: []types.SwapAmountInRoute{
+					{
+						PoolId:        fooBarPoolId,
+						TokenOutDenom: bar,
+					},
+				},
+				TokenInAmount: twentyFiveBaseUnitsAmount,
+			},
+		}
+
+		defaultTwoHopRoutes = []types.SwapAmountInRoute{
+			{
+				PoolId:        fooBarPoolId,
+				TokenOutDenom: bar,
+			},
+			{
+				PoolId:        barBazPoolId,
+				TokenOutDenom: baz,
+			},
+		}
+
+		defaultSingleRouteTwoHops = types.SwapAmountInSplitRoute{
+			Pools:         defaultTwoHopRoutes,
+			TokenInAmount: twentyFiveBaseUnitsAmount,
+		}
+
+		defaultSingleRouteThreeHops = types.SwapAmountInSplitRoute{
+			Pools: []types.SwapAmountInRoute{
+				{
+					PoolId:        fooBarPoolId,
+					TokenOutDenom: bar,
+				},
+				{
+					PoolId:        barUosmoPoolId,
+					TokenOutDenom: uosmo,
+				},
+				{
+					PoolId:        bazUosmoPoolId,
+					TokenOutDenom: baz,
+				},
+			},
+			TokenInAmount: sdk.NewInt(twentyFiveBaseUnitsAmount.Int64() * 3),
+		}
+
+		priceImpactThreshold = sdk.NewInt(97866545)
+	)
+
+	tests := map[string]struct {
+		isInvalidSender   bool
+		routes            []types.SwapAmountInSplitRoute
+		tokenInDenom      string
+		tokenOutMinAmount sdk.Int
+
+		// This is an estimate because we do not care about
+		// testing exact values at this level of abstraction.
+		// The math should be tested per-module.
+		// We keep this assertion to make sure that the
+		// actual result is within a reasonable range.
+		expectedTokenOutEstimate sdk.Int
+
+		expectError error
+	}{
+		"valid solo route one hop": {
+			routes:            defaultSingleRouteOneHop,
+			tokenInDenom:      foo,
+			tokenOutMinAmount: sdk.OneInt(),
+
+			expectedTokenOutEstimate: twentyFiveBaseUnitsAmount,
+		},
+		"valid solo route multi hop": {
+			routes:            []types.SwapAmountInSplitRoute{defaultSingleRouteTwoHops},
+			tokenInDenom:      foo,
+			tokenOutMinAmount: sdk.OneInt(),
```

Comment:
asking for my own learning, what is significant about one here? Can this not be 0?

Reply:
It can, just small value that we know well not trigger the price impact protection